Fixes #2424 Enhanced Tracking Protection

keianhzo opened the pull request:
Fixes #2424 Enhanced Tracking Protection

keianhzo commented:
There are a few common issues with the WebXR settings PR:
- Dialog not closed when switching windows
- Dialog not closed when clicking on the button again
- Hide buttons in the content feed.

I'll fix those when merging the pending PRs that already address those.

bluemarvin commented on app/src/common/shared/org/mozilla/vrbrowser/browser/engine/EngineProvider.kt:
```diff
@@ -25,7 +25,11 @@ object EngineProvider {
+                            ContentBlocking.AntiTracking.AD or
+                            ContentBlocking.AntiTracking.SOCIAL or
+                            ContentBlocking.AntiTracking.ANALYTIC)
+                    .enhancedTrackingProtectionLevel(SettingsStore.getInstance(context).trackingProtectionLevel)
```
I think we only want one or the other set at one time?

droeh replied:
These are independent, actually. `antiTracking` controls what *content* gets blocked when `GeckoSessionSettings.useTrackingProtection` is true, while `enhancedTrackingProtectionLevel` controls what *cookies* get blocked when our cookie behavior is `ACCEPT_NON_TRACKERS`.

bluemarvin replied:
The way I understood what was explained to me is that we only want one active at a time. If we are blocking cookies we don't want to be blocking content as that is the purpose of ETP, it just blocks the tracking cookies and the content is all loaded which reduce breakage, where content blocking is more strict but also causes more breakage.

bluemarvin replied:
> These are independent, actually. `antiTracking` controls what _content_ gets blocked when `GeckoSessionSettings.useTrackingProtection` is true, while `enhancedTrackingProtectionLevel` controls what _cookies_ get blocked when our cookie behavior is `ACCEPT_NON_TRACKERS`.

@droeh I guess I'm still confused, I don't see `GeckoSessionSettings.useTrackingProtection()` except when the session is created. Does this mean it never gets turned off?

droeh replied:
Sorry for the slow reponse, I'm not used to getting work mail from github!
There are cases where you might want to leave ETP and TP both enabled, for example you may want to enable TP only for one category and ETP for the rest. In general they are only exposed in UI mutually exclusively, though (eg as "Standard" and "Strict" in desktop).

And yeah, I missed that, but you definitely need to update `GeckoSessionSettings.useTrackingProtection()` to control TP, and do it for every session. (Yes, that sucks, it's on the list of things to change in the API.)

I'll keep an eye on github mail and respond quicker if you have any more questions.

bluemarvin commented on app/src/common/shared/org/mozilla/vrbrowser/browser/engine/Session.java:
```diff
@@ -1239,6 +1231,25 @@ public void onContentBlocked(@NonNull final GeckoSession session, @NonNull final
+        if ((event.getAntiTrackingCategory() & ContentBlocking.AntiTracking.SOCIAL) != 0) {
+            Log.i(LOGTAG, "Loading Social: " + event.uri);
+        }
+    }
```
These need to be `Log.d` so they get stripped out of release builds or we will get rejected from HTC store.

bluemarvin commented on app/src/common/shared/org/mozilla/vrbrowser/browser/tracking/TrackingProtectionPolicy.java:
```diff
@@ -0,0 +1,133 @@
+    public @interface ETP_Level {}
+    public static final int ETP_NONE = ContentBlocking.EtpLevel.NONE;
+    public static final int ETP_DEFAULT = ContentBlocking.EtpLevel.DEFAULT;
+    public static final int ETP_STRICT = ContentBlocking.EtpLevel.STRICT;
```
I'm not sure why we need to do this.

bluemarvin commented on app/src/common/shared/org/mozilla/vrbrowser/browser/tracking/TrackingProtectionPolicy.java:
```diff
@@ -0,0 +1,133 @@
+        RECOMMENDED(AD.value + ANALYTICS.value + SOCIAL.value +
+                TEST.value + MOZILLA_SOCIAL.value + CRYPTOMINING.value),
+        STRICT(RECOMMENDED.value + SCRIPTS_AND_SUB_RESOURCES.value +
+                FINGERPRINTING.value);
```
And this, should we just use the values in GeckoView so if they change we don't need to keep them synchronized?

droeh commented on app/src/common/shared/org/mozilla/vrbrowser/browser/tracking/TrackingProtectionPolicy.java:
```diff
@@ -0,0 +1,133 @@
+        CRYPTOMINING(1 << 6),
+        FINGERPRINTING(1 << 7),
+        MOZILLA_SOCIAL(1 << 8),
+        SCRIPTS_AND_SUB_RESOURCES(1 << 9999),
```
`1 << 9999` is very alarming unless Java does something cute here that I'm not aware of.

bluemarvin replied:
yikes, that is bad, I assume it is just zero which is probably not what they intended. I'm not sure we even want all this.

bluemarvin commented:
@keianhzo droeh says this is the site to use for testing: https://senglehardt.com/test/trackingprotection/test_pages/

keianhzo commented:
@bluemarvin @droeh I've updated and removed all redundant code. I've been testing with the site above and I think the results are good, also are the sames as Fenix.

bluemarvin commented:
Seems to work my tests. It would be nice if we could do a v2 where the shield changed colors when it actually blocks some thing and then clicking on that would give more details on what was blocked.

bluemarvin commented on app/src/common/shared/org/mozilla/vrbrowser/browser/engine/EngineProvider.kt:
```diff
@@ -25,7 +25,11 @@ object EngineProvider {
+                    .antiTracking(
+                            ContentBlocking.AntiTracking.AD or
+                            ContentBlocking.AntiTracking.SOCIAL or
+                            ContentBlocking.AntiTracking.ANALYTIC)
```
I think we should be using the level from the preferences here. I assume it gets over-written later but we should be starting with the correct anti-tracking values I think.

keianhzo replied:
Makes sense, updated

bluemarvin commented:
I wonder if we should even show the shield if tracking protection is turned off.

keianhzo commented:
@bluemarvin Regarding:

> But it says tracking protection is still on and cookies are still getting blocked:

I see the same behavior in Fenix, maybe @droeh can clarify.

And regarding:

> I wonder if we should even show the shield if tracking protection is turned off.

In Desktop/Fenix you can never completely turn ETP off so the icon is always visible. Maybe if ETP is **Off** we could show the disabled icon, change the bubble text to something like "ETP is disabled. Go to setting to turn it on" and make the button to take you to the ETP settings. It's just weird that in desktop and Fenix is always visible and we hide it. Also it is faster to go to the setting from there if you want to re-enable again.

keianhzo commented:
@bluemarvin updated with hiding the tracking protection icon if disabled.

bluemarvin commented:
https://github.com/MozillaReality/FirefoxReality/issues/3053
